### PR TITLE
MTDSA-5253 Amend Validator & spec

### PR DESCRIPTION
**Pull request by zachhall1234**

### CODE REVIEW

File: app/v1/controllers/requestParsers/validators/AmendValidator.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    List(
+      PeriodDataPositiveAmountValidation.validate(data.body, "deductionAmount", RuleDeductionAmountError),
+      PeriodDataPositiveAmountValidation.validate(data.body, "costOfMaterials", RuleCostOfMaterialsError),
+      PeriodDataPositiveAmountValidation.validate(data.body, "grossAmountPaid", RuleGrossAmountError),
+      PeriodDataDeductionDateValidation.validate(data.body, "deductionFromDate", DeductionFromDateFormatError),
+      PeriodDataDeductionDateValidation.validate(data.body, "deductionToDate", DeductionToDateFormatError),
+      DateValidation.validate(FromDateFormatError)(req.fromDate),
+      DateValidation.validate(ToDateFormatError)(req.toDate),
+      ToBeforeFromDateValidation.validate(req.fromDate, req.toDate, RuleDateRangeInvalidError)
+    )
```
Comment (ian-ralphs) [inline app/v1/controllers/requestParsers/validators/AmendValidator.scala:54]: need to add rule date validations for start of tax year and end of tax year - will require extra tests too.

File: app/v1/controllers/requestParsers/validators/validations/BodyTaxYearValidation.scala [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+object BodyTaxYearValidation {
+
+  def validate(date: String, fieldName: String, error: MtdError): List[MtdError] = {
+    fieldName match {
+      case "fromDate" => validateFromDate(error)(date)
+      case "toDate" => validateToDate(error)(date)
+    }
+  }
+
+  val fromDateFormat = "[0-9][0-9][0-9][0-9]-[0][4]-[0][6]"
```
Comment (ian-ralphs) [inline app/v1/controllers/requestParsers/validators/validations/BodyTaxYearValidation.scala:30]: Regex more eligantly put as [0-9]{4}-04-06
4 numbers in range 0-9 then an exact match on -04-06

File: test/v1/fixtures/CreateRequestFixtures.scala [hunk 3]
```diff
@@ -467,6 +467,60 @@ object CreateRequestFixtures {
       |""".stripMargin
   }
 
+  val invalidFromDateForTaxYear: JsValue = Json.parse {
```
Comment (ian-ralphs) [inline test/v1/fixtures/CreateRequestFixtures.scala:470]: I think these should be in the Amend fixtures file

File: app/v1/controllers/requestParsers/validators/AmendValidator.scala [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  private def businessRuleValidator: AmendRawData => List[List[MtdError]] = { data =>
+    val req = data.body.as[AmendRequest]
+
+    List(
```
Comment (ian-ralphs) [inline app/v1/controllers/requestParsers/validators/AmendValidator.scala:61]: I'm think we should just have this as 1 call which will take both the from and to dates plus an optional "single year" flag. The from date and to date are validated as you currently are then if the single year flag is specified a further check will test if the toDate year is 1 more than the from date year. The error codes don't need to be passed in as the function is specific to the task of validating a tax year. You can probably drop the "Body" bit from the name too.

File: app/v1/controllers/requestParsers/validators/validations/TaxYearDatesValidation.scala [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import java.time.format.DateTimeFormatter
+import v1.models.errors.{MtdError, RuleDateRangeInvalidError, RuleFromDateError, RuleToDateError}
+
+object TaxYearDatesValidation {
+
+  def validate(fromDate: String, toDate: String, numYears: Option[Int]): List[MtdError] = {
+    val dateFormatErrs = validateFromDate(fromDate) ++
+    validateToDate(toDate)
+    if(numYears.isDefined & dateFormatErrs.isEmpty)
+    {validateYears(numYears.get, fromDate, toDate)}
```
Comment (ian-ralphs) [inline app/v1/controllers/requestParsers/validators/validations/TaxYearDatesValidation.scala:29]: curly braces around a single code line are not required

Comment (zachhall1234) [reply app/v1/controllers/requestParsers/validators/validations/TaxYearDatesValidation.scala]: Removed

**Pull request merged**